### PR TITLE
Generate `replaces` automatically in builder-integration-test

### LINKED ISSUE #11607: [release] builder integration test succeeds on main but fails on release prep
**Describe the bug**

When a new module was added since the last release, the `Builder - Integration tests` github action succeeds on main but fails on the [release prep PR](https://github.com/open-telemetry/opentelemetry-collector/actions/runs/11687864326/job/32546947029?pr=11602#step:4:24).

**What did you expect to see?**

CI should be able to catch this issue immediately and appropriate replace statements should be required in the codebase.

**What did you see instead?**

The release process fails.

---

Is it possible to detect and require a solution to this problem at the time the module was added? What new tests or changes to existing tests are necessary to ensure releases are not blocked by this in the future?

### PR DESCRIPTION
**Pull request by jade-guiton-dd**
#### Description

This PR removes the manually maintained `replace` statements in `cmd/builder/test/core.builder.yaml` and generates them automatically in `cmd/builder/test/test.sh`. This ensures that the builder integration test properly uses the local version of core collector modules, even if a `replace` statement is forgotten. This is especially important in release PRs where dependencies have not-yet-valid version numbers (see tracking issue for context).

#### Note

I believe a better implementation of this long-term may be to commit a `go.work` file to the repository as the single source of truth for `replace`s, and let the `go` commands run by `ocb` take it into account automatically. (This would also avoid needing to maintain the `replace`s in every `go.mod`, which can be annoying when `make crosslink` is not sufficient.) But as there were [objections the last time this was discussed](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21972#issuecomment-1549173972), I decided to leave this as a future discussion.

#### Link to tracking issue
Fixes #11607

#### Testing
We can simulate a Release PR by setting the `exporter` import in `debugexporter` to an invalid version like `0.999.0`. With the new script, this causes no problems, showing that the build uses the local version without fetching from the proxy. Filtering the `exporter` module from the replace statements by adding this to the script:
```bash
core_mods=$(echo "$core_mods" | grep -v "/exporter$")
```
makes `ocb` output the expected `unknown revision` error.

### CODE REVIEW

**Review by mx-psi — COMMENTED** [review]
Should we keep the replaces in the code and have a CI check that fails if they are not generated correctly? Similar to what we do with code generation elsewhere

**Comment by jade-guiton-dd** [comment]
> Should we keep the replaces in the YAML and have a CI check that fails if they are not generated correctly? Similar to what we do with code generation elsewhere

We could do that, but I feel like that would only be useful if people are planning to use the integration test's builder config manually, outside of the integration test script. Considering the config file already requires calling `envsubst` on it before being usable, I doubt this is a real use case.

It's also surprisingly difficult to compute the proper list of required `replace` statements like we have now: you need to include not only the components included in the YAML, but also the modules added in automatically by `ocb` (not sure if there is a machine-readable list of that somewhere), AND all their transitive dependencies. I tried doing that at first, but I ended up giving up in favor of just including every single core module.

**Comment by mx-psi** [comment]
> We could do that, but I feel like that would only be useful if people are planning to use the integration test's builder config manually, outside of the integration test script. Considering the config file already requires calling envsubst on it before being usable, I doubt this is a real use case.

The usefulness, IMO, comes from being able to see how changes in the script affect this step of it. But if the list is always going to be all modules, then it feels like this is reasonable to me

File: cmd/builder/test/test.sh [hunk 1]
```diff
@@ -113,6 +113,14 @@ tests="core"
 base=$(mktemp -d)
 echo "Running the tests in ${base}"
 
+replaces="$base/replaces"
+core_mods=$(cd ../.. && find . -type f -name "go.mod" -exec dirname {} \; | sort | sed 's/.//')
```
[the remark this answers is not in the document]
Comment (jade-guiton-dd) [reply cmd/builder/test/test.sh:117]: The sort is just to get alphabetical order, which isn't strictly necessary but makes the output easier to debug if necessary. `sed` just removes the initial dot from the paths; this is useful later when concatenating with "go.opentelemetry.io/collector" and "${WORKSPACE_DIR}" below. I'll add a comment about it.

**Review by evan-bradley — APPROVED** [review]

**Review by mx-psi — APPROVED** [review]

**Review by songy23 — COMMENTED** [review]
This caused failure in releases https://github.com/open-telemetry/opentelemetry-collector/actions/runs/12358329683/job/34488569900?pr=11904 seems the replace isn't working properly

**Pull request merged**